sql-plugin/src/main/330+/scala/com/nvidia/spark/rapids/shims/v2/GpuParquetScan.scala

```diff
@@ -0,0 +1,203 @@
+import org.apache.spark.sql.execution.datasources.{AggregatePushDownUtils, DataSourceUtils, PartitionedFile, PartitioningAwareFileIndex}
+import org.apache.spark.sql.execution.datasources.parquet.{ParquetFooterReader, ParquetOptions}
+import org.apache.spark.sql.execution.datasources.parquet.rapids.shims.v2.GpuParquetUtils
+import org.apache.spark.sql.execution.datasources.v2.{FilePartitionReaderFactory, FileScan}
+import org.apache.spark.sql.internal.SQLConf
+import org.apache.spark.sql.sources.Filter
+import org.apache.spark.sql.types.StructType
+import org.apache.spark.sql.util.CaseInsensitiveStringMap
+import org.apache.spark.sql.vectorized.ColumnarBatch
+import org.apache.spark.util.SerializableConfiguration
+
+/**
+ * Users should make sure that `aggregation` is valid when using this factory.
+ */
+case class GpuParquetPartitionReaderFactoryForAggregationPushdown(
+    @transient sqlConf: SQLConf,
+    broadcastedConf: Broadcast[SerializableConfiguration],
+    dataSchema: StructType,
+    readDataSchema: StructType,
+    partitionSchema: StructType,
+    filters: Array[Filter],
+    aggregation: Aggregation,
+    parquetOptions: ParquetOptions,
+    @transient rapidsConf: RapidsConf,
+  metrics: Map[String, GpuMetric]) extends FilePartitionReaderFactory with Arm with Logging {
+  private val datetimeRebaseModeInRead = parquetOptions.datetimeRebaseModeInRead
+
+  private def getFooter(file: PartitionedFile): ParquetMetadata = {
+    val conf = broadcastedConf.value.value
+    val filePath = new Path(new URI(file.filePath))
+
+    // For aggregate push down, we will get max/min/count from footer statistics.
+    ParquetFooterReader.readFooter(conf, filePath, NO_FILTER)
+  }
+
+  private def getDatetimeRebaseSpec(
+    footerFileMetaData: FileMetaData): RebaseSpec = {
+    DataSourceUtils.datetimeRebaseSpec(
+      footerFileMetaData.getKeyValueMetaData.get,
+      datetimeRebaseModeInRead)
+  }
+
+  override def supportColumnarReads(partition: InputPartition): Boolean = true
+
+  override def buildReader(partitionedFile: PartitionedFile): PartitionReader[InternalRow] = {
+    throw new IllegalStateException("GPU column parser called to read rows")
+  }
+
+  override def buildColumnarReader(
+    partitionedFile: PartitionedFile): PartitionReader[ColumnarBatch] = {
+    new PartitionReader[ColumnarBatch] {
+      private var hasNext = true
+      private val batch: ColumnarBatch = {
+        val footer = getFooter(partitionedFile)
+        if (footer != null && footer.getBlocks.size > 0) {
+          val row = GpuParquetUtils.createAggInternalRowFromFooter(
+            footer,
+            partitionedFile.filePath,
+            dataSchema,
+            partitionSchema,
+            aggregation,
+            readDataSchema,
+            partitionedFile.partitionValues,
+            getDatetimeRebaseSpec(footer.getFileMetaData)
+          )
+
+          new GpuRowToColumnConverter(readDataSchema).convertBatch(Array(row), readDataSchema)
```

Comment:
If I'm reading this code correctly, it seems like we're going out of our way to translate a Parquet scan to the GPU even though at planning time we know that this will just be a CPU-only read of the footer and agg statistics from it.  The GPU "implementation" is just calling the underlying Spark code and then following it up with a RowToColumnar transition.  Why not simply have the tagging code _not_ replace a Parquet scan when we know that the Parquet scan will simply be reading aggregation values from the footer?  Then the RowToColumnar will be put into the plan automatically when necessary.

This also can avoid performance issues where it looks like the GPU is performing the Parquet scan (when it really isn't) and we convert it from row to columnar here internally, but then in the query plan this node is followed by something we cannot do on the GPU, so then we follow up this internal RowToColumn conversion with a ColumnarToRow conversion in the query plan to continue execution on the CPU when we should have remained on the CPU in row mode all along.

Reply:
Yes, you are right. Fallback to CPU is the reasonable solution. I will close this PR and reopen a new one.